npep/npep-136.md

```diff
@@ -0,0 +1,102 @@
+# NPEP-136: DeveloperNetworkPolicy
+
+* Issue: [#136](https://github.com/kubernetes-sigs/network-policy-api/issues/136)
+* Status: Provisional
+
+## TLDR
+
+I want to have the next generation of NetworkPolicies for Application Developers as an addition to Admin-focused APIs.
+
+## Goals
+
+Add the missing part of "NetworkPolicy v2": ANP API is the cluster-scoped and is created for cluster admins, 
+DeveloperNetworkPolicy (DNP for short) is namespace-scoped and is created for application developer/namespace owners.
+
+## Non-Goals
+
+This is not the next version of an existing NetworkPolicy API, therefore it doesn't need to be compatible with
+NetworkPolicy, but it needs to be fully interoperable (i.e. have well-defined behaviour when used with NetworkPolicy).
+
+## Introduction
+
+Mostly coming from https://docs.google.com/document/d/10L9_ZV7nmabObFSKs4rR_t8ATtOTBPwsZKKSB6GFYp0/edit
+
+We want to build a new API, starting from scratch, and not building on top of the NetworkPolicy. Therefore, this NPEP
+will define the developer-focused user stories from scratch. There are some general properties that are desired for the new API
+(that we mostly learned thanks to NetworkPolicy)
+
+1. It needs to be extensible
+
+    We keep running into the problem in NetworkPolicy that we want to add new “stuff”, but the new stuff would be 
+    invisible to old implementations. We need to fix this in DNP.
+2. It needs to be explicit
+
+   - It should not have “implicit deny on first policy”. Denies, like accepts, should be explicit.
+   This implies that you should be able to create a policy that says “allow X”, where if X wasn’t already denied 
+   (eg by BaselineAdminNetworkPolicy) then the “allow X” would just be a no-op (rather than causing other traffic to be denied like with NPv1).
+   - It should not have the automatic “connections from the pod’s node IP are always accepted” exception.
+   (but you shouldn’t have to manually add that rule either… we probably need to make probes work differently)
+   TODO: add link to the probes discussion
+   - It should not have tricky “arrays of objects where multiple fields can be set but none are required” 
+   such that adding or removing a single “-” drastically changes the meaning of the policy.
+   - “Present but empty” / “Present but zero length” should not mean something different from “not present” / nil. 
+   Likewise, the difference between an array of length 0 and an array of length 1 should be the same as the difference 
+   between an array of length 1 and an array of length 2.
+
+DeveloperNetworkPolicy should work well together with ANP and BANP as a part of the same API.
+
+## User-Stories/Use-Cases
+
+### User Story 1: Isolate namespace
+
+As a namespace owner, I want to make sure my pods can only access other pods from the same namespace and nothing else.
+I also want to restrict incoming connections from outside the namespace.
```

Comment:
The ingress and egress cases are different stories.

You need ingress deny rules if you want to protect your pods from other pods in the cluster. You don't control other users' pods, so you can't prevent them from _trying_ to connect to your pods, so you need ingress deny rules to block their connection attempts.

But the same doesn't apply to egress deny rules. If you don't want your pod to connect to pods in other namespaces then you can just _not connect to pods in other namespaces_. The only reason you'd need an egress deny rule is if you were worried that your pods might try to connect to other namespaces even though they weren't configured to. So egress deny rules are to protect you from hackers (or bugs), not to protect you from malicious other users within the cluster like ingress deny rules.

Reply:
I always have this concept in mind when talking about egress namespace rules, but I think it is not written anywhere, so may be a good chance to clarify that.